File: java/MyCd.java
import java.util.Arrays;

public class MyCd {

    public static String myCd(String fromDir, String toDir) {
        String path = fromDir;

        if (isAbsolutePath(toDir)) {
            path = "/";
        }

        for (String d : toDir.split("/")) {
            if (d.equals("..")) {
                path = removeLastDir(path);
            } else if (dirIsValid(d)) {
                path += "/" + d;
            } else if (!staySameDirectory(d)) {
                return getErrorMessage(toDir);
            }
        }

        return normalizePath(path);
    }

    private static boolean isAbsolutePath(String path) {
        return path.startsWith("/");
    }

    private static boolean staySameDirectory(String d) {
        return d.equals(".") || d.equals("");
    }

    private static String getErrorMessage(String toDir) {
        return toDir + ": No such file or directory";
    }

    private static boolean dirIsValid(String d) {
        return d.matches("[A-Za-z0-9]+");
    }

    private static String removeLastDir(String path) {
        String[] temp = path.split("/");
        if (temp.length > 0) {
            temp = Arrays.copyOf(temp, temp.length-1);
            return String.join("/", temp);
        }
        return path;
    }

    private static String normalizePath(String path) {
        path = path.replaceAll("/+", "/");
        return (path.startsWith("/") ? "" : "/") + path;
    }

}

File: python/mycd.py
import re

def mycd(fromdir, todir):
    path = fromdir
    if todir[0] == "/":
            path = todir
    else:
        for d in todir.split("/"):
            if d in [".", ""]:
                pass
            elif d == "..":
                temp = path.split("/")
                temp.pop()
                path = "/".join(temp)
            elif d.isalnum():
                path += "/"+d
            else:
                return f"{todir}: No such file or directory"
    return tidy_path(path)

def tidy_path(path):
    path = re.sub("\/+", "/", path)
    return ("" if path.startswith("/") else "/") + path